r/R/dataset.R

```diff
@@ -0,0 +1,191 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#' Open a multi-file dataset
+#'
+#' @param path String path to a directory containing the data files
+#' @param schema [Schema] for the dataset. If `NULL` (the default), the schema
+#' will be inferred from the files
+#' @param partition One of
+#'   * A `Schema`, in which case the file paths relative to `path` will be
+#'    parsed, and path segments will be matched with the schema fields. For
+#'    example, `schema(year = int16(), month = int8())` would create partitions
+#'    for file paths like "2019/01/file.parquet", "2019/02/file.parquet", etc.
+#'   * A `HivePartitionScheme`, as returned by [hive_partition()]
+#'   * `NULL`, the default, for no partitioning
+#' @param ... additional arguments passed to `DataSourceDiscovery$create()`
+#' @return A [Dataset] R6 object. Use `dplyr` methods on it to query the data,
+#' or call `$NewScan()` to construct a query directly.
+#' @export
+#' @include arrow-package.R
+open_dataset <- function (path, schema = NULL, partition = NULL, ...) {
```

Comment:
`open_file_dataset` maybe?

Reply:
Why would that be better?

Reply:
because there might be other type of dataset (datasources), e.g. sql/odbc, flight...

Reply:
Whenever we get there, I'll extend this function to take other input sources.